src/clj/puppetlabs/services/ca/certificate_authority_core.clj

```diff
@@ -56,11 +59,34 @@
       ;; Respond to all CSR validation failures with a 400
       (middleware-utils/plain-response 400 msg))))
 
-(defn handle-get-certificate-revocation-list
-  [{:keys [cacrl]}]
-  (-> (ca/get-certificate-revocation-list cacrl)
-      (rr/response)
-      (rr/content-type "text/plain")))
+  (schema/defn format-http-date :- org.joda.time.DateTime
+    "Formats a http-date into joda time"
+    [http-date :- schema/Str]
+    (time-format/parse
+      (time-format/formatters :rfc822)
+      (string/replace http-date #"GMT" "+0000")))
+
+  (defn handle-get-certificate-revocation-list
+    "Always return the crl if no 'if-modified-since' header is provided or if that header is not in correct http-date format.
+    If the header is present and correctly formatted, only return the crl if the master cacrl is newer than the agent crl."
+    [request {:keys [cacrl]}]
+    (let [agent-crl-last-modified-val (rr/get-header request "If-Modified-Since")
+          agent-crl-last-modified-date-time (if agent-crl-last-modified-val
+                                              (try
+                                                (format-http-date agent-crl-last-modified-val)
+                                                (catch IllegalArgumentException e
+                                                  :invalid-http-date))
+                                              :unknown)
```

Comment:
Personally I would move everything starting with the `if` into that function you made, so that the `let` binding here is just naming the result of calling that function, but I'd appreciate the thoughts of someone more experienced with Clojure. Still trying to get a feel for what is considered readable.

If you do that, I would definitely just return nil for the unhappy paths. Then this becomes "Ask for agent modified time, either get it or get nothing, handle those two cases."

Reply:
^ I agree with second part.  I considered both options.  I was thinking...that if this method were to be called in the future that the error handling desired may be different (so maybe it should exist w/out error handling packaged in).  Also, it would have to accept either a nil or string argument, so I couldn't use schema/str.  And not sure what clojure thinks about this ambiguity.

Reply:
i think moving it makes sense

it could go either way, but handle-get-certificate-revocation-list is long enough that it makes sense to extract that

you could always have two functions, one that does the parsing and one that does error handling, but someone who needs to distinguish that could always do that later

Reply:
^ Updated.  Thanks for the great advice everyone.